src/PosthogAnalytics.ts

```diff
@@ -0,0 +1,343 @@
+ * - If both flags are false, events are not sent.
+*/
+
+interface IEvent {
+    // The event name that will be used by PostHog.
+    // TODO: standard format (camel case? snake? UpperCase?)
+    eventName: string;
+
+    // The properties of the event that will be stored in PostHog.
+    properties: {};
+}
+
+export enum Anonymity {
+    Disabled,
+    Anonymous,
+    Pseudonymous
+}
+
+// If an event extends IPseudonymousEvent, the event contains pseudonymous data
+// that won't be sent unless the user has explicitly consented to pseudonymous tracking.
+// For example, it might contain hashed user IDs or room IDs.
+// Such events will be automatically dropped if PosthogAnalytics.anonymity isn't set to Pseudonymous.
+export interface IPseudonymousEvent extends IEvent {}
+
+// If an event extends IAnonymousEvent, the event strictly contains *only* anonymous data;
+// i.e. no identifiers that can be associated with the user.
+export interface IAnonymousEvent extends IEvent {}
+
+export interface IRoomEvent extends IPseudonymousEvent {
+    hashedRoomId: string;
+}
+
+interface IPageView extends IAnonymousEvent {
+    eventName: "$pageview";
+    properties: {
+        durationMs?: number;
+        screen?: string;
+    };
+}
+
+export interface IWelcomeScreenLoad extends IAnonymousEvent {
+    eventName: "welcome_screen_load";
+}
+
+const hashHex = async (input: string): Promise<string> => {
+    // on os x (e.g. if you want to know the sha-256 of your own matrix ID so you can look it up):
+    // echo -n <input> | shasum -a 256
+    const buf = new TextEncoder().encode(input);
+    const digestBuf = await window.crypto.subtle.digest("sha-256", buf);
+    return [...new Uint8Array(digestBuf)].map((b: number) => b.toString(16).padStart(2, "0")).join("");
+};
+
+const knownScreens = new Set([
+    "register", "login", "forgot_password", "soft_logout", "new", "settings", "welcome", "home", "start", "directory",
+    "start_sso", "start_cas", "groups", "complete_security", "post_registration", "room", "user", "group",
+]);
+
+export async function getRedactedCurrentLocation(origin: string, hash: string, pathname: string, anonymity: Anonymity) {
+    // Redact PII from the current location.
+    // If anonymous is true, redact entirely, if false, substitute it with a hash.
+    // For known screens, assumes a URL structure of /<screen name>/might/be/pii
+    if (origin.startsWith('file://')) {
+        pathname = "/<redacted_file_scheme_url>/";
+    }
+
+    let hashStr;
+    if (hash == "") {
+        hashStr = "";
+    } else {
+        let [_, screen, ...parts] = hash.split("/");
+
+        if (!knownScreens.has(screen)) {
+            screen = "<redacted_screen_name>";
+        }
+
+        for (let i = 0; i < parts.length; i++) {
+            parts[i] = anonymity === Anonymity.Anonymous ? `<redacted>` : await hashHex(parts[i]);
+        }
+
+        hashStr = `${_}/${screen}/${parts.join("/")}`;
```

Comment:
If we use the identifier `_` its because we don't want to use it, but here were using it and without it having a name I have no idea what it is

Reply:
👍 this function was lifted from https://github.com/matrix-org/matrix-react-sdk/blame/b8781b75523350c987bd4e8c24535331f2dd3f01/src/CountlyAnalytics.ts#L313